src/snowflake/snowpark/modin/plugin/_internal/index.py

```diff
@@ -0,0 +1,1112 @@
+# Version 2.0.
+
+"""Module houses ``Index`` class, that is distributed version of ``pandas.Index``."""
+
+from __future__ import annotations
+
+from typing import Any, Callable, Hashable, Iterator, Literal
+
+import numpy as np
+import pandas as native_pd
+from pandas._typing import ArrayLike, DtypeObj, NaPosition, Self
+from pandas.core.arrays import ExtensionArray
+from pandas.core.dtypes.base import ExtensionDtype
+from pandas.core.indexes.frozen import FrozenList
+
+from snowflake.snowpark.modin.pandas.api.extensions import register_pd_accessor
+from snowflake.snowpark.modin.pandas.utils import try_convert_index_to_native
+from snowflake.snowpark.modin.plugin.utils.warning_message import WarningMessage
+
+
+@register_pd_accessor("Index")
+class Index:
+    """
+    Immutable sequence used for indexing and alignment.
+
+    The basic object storing axis labels for all pandas objects.
+
+    Parameters
+    ----------
+    data : array-like (1-dimensional)
+    dtype : str, numpy.dtype, or ExtensionDtype, optional
+        Data type for the output Index. If not specified, this will be
+        inferred from `data`.
+        See the :ref:`user guide <basics.dtypes>` for more usages.
+    copy : bool, default False
+        Copy input data.
+    name : object
+        Name to be stored in the index.
+    tupleize_cols : bool (default: True)
+        When True, attempt to create a MultiIndex if possible.
+
+    Notes
+    -----
+    An Index instance can **only** contain hashable objects.
+    An Index instance *can not* hold numpy float16 dtype.
+
+    Examples
+    --------
+    >>> pd.Index([1, 2, 3])
+    Index([1, 2, 3], dtype='int64')
+
+    >>> pd.Index(list('abc'))
+    Index(['a', 'b', 'c'], dtype='object')
+
+    >>> pd.Index([1, 2, 3], dtype="uint8")
+    Index([1, 2, 3], dtype='uint8')
+    """
+
+    # same fields as native pandas index constructor
+    def __init__(
+        self,
+        # Any should be replaced with SnowflakeQueryCompiler when possible (linter won't allow it now)
+        data: ArrayLike | Any = None,
+        dtype: str | np.dtype | ExtensionDtype | None = None,
+        copy: bool = False,
+        name: object = None,
+        tupleize_cols: bool = True,
+    ) -> None:
+        from snowflake.snowpark.modin.plugin.compiler.snowflake_query_compiler import (
+            SnowflakeQueryCompiler,
+        )
+
+        # TODO: SNOW-1359041: Switch to lazy index implementation
+        if isinstance(data, native_pd.Index):
+            self._index = data
+        elif isinstance(data, Index):
+            self._index = data.to_pandas()
+        elif isinstance(data, SnowflakeQueryCompiler):
+            self._index = data._modin_frame.index_columns_index
+        else:
+            self._index = native_pd.Index(
+                data=data,
+                dtype=dtype,
+                copy=copy,
+                name=name,
+                tupleize_cols=tupleize_cols,
+            )
+        self._name = name
```

Comment:
so we can get or set names both in self._name and self._index.name? Let's introduce `_name` later?